src/app/clusters/webrtc-transport-provider-server/webrtc-transport-provider-server.cpp

```diff
@@ -0,0 +1,495 @@
+    }
+
+    // Store the WebRTCSessionStruct in the CurrentSessions
+    AddOrUpdateSession(outSession);
+
+    // Send the response
+    Commands::ProvideOfferResponse::Type resp;
+    resp.webRTCSessionID = outSession.id;
+    resp.audioStreamID.SetValue(outSession.videoStreamID);
+    resp.audioStreamID.SetValue(outSession.audioStreamID);
+
+    ctx.mCommandHandler.AddResponse(ctx.mRequestPath, resp);
+}
+
+void WebRTCTransportProviderServer::HandleProvideAnswer(HandlerContext & ctx, const Commands::ProvideAnswer::DecodableType & req)
+{
+    // Extract command fields from the request.
+    uint16_t sessionId = req.webRTCSessionID;
+    auto sdpAnswer     = req.sdp;
+
+    NodeId peerNodeId           = GetNodeIdFromCtx(ctx.mCommandHandler);
+    FabricIndex peerFabricIndex = ctx.mCommandHandler.GetAccessingFabricIndex();
+
+    WebRTCSessionStruct * existingSession = FindSession(sessionId);
+
+    if (!existingSession)
+    {
+        // If WebRTCSessionID is not null and does not match a value in current sessions.
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Protocols::InteractionModel::Status::ConstraintError);
+        return;
+    }
+
+    // Also check that the existing session belongs to the same PeerNodeID / Fabric
+    // If it doesn’t match, respond with CONSTRAINT_ERROR
+    if (peerNodeId != existingSession->peerNodeID || peerFabricIndex != existingSession->peerFabricIndex)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Protocols::InteractionModel::Status::ConstraintError);
+        return;
+    }
+
+    Status status  = Protocols::InteractionModel::Status::Success;
+    CHIP_ERROR err = mDelegate->HandleProvideAnswer(sessionId, sdpAnswer);
+    if (err != CHIP_NO_ERROR)
+    {
+        status = StatusIB(err).mStatus;
+    }
+
+    ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+}
+
+void WebRTCTransportProviderServer::HandleProvideICECandidates(HandlerContext & ctx,
+                                                               const Commands::ProvideICECandidates::DecodableType & req)
+{
+    // Extract command fields from the request.
+    uint16_t sessionId = req.webRTCSessionID;
+
+    NodeId peerNodeId           = GetNodeIdFromCtx(ctx.mCommandHandler);
+    FabricIndex peerFabricIndex = ctx.mCommandHandler.GetAccessingFabricIndex();
+
+    std::vector<std::string> candidates;
+    auto iter = req.ICECandidates.begin();
+    while (iter.Next())
+    {
+        // Get current CharSpan
+        const chip::CharSpan & candidateSpan = iter.GetValue();
+
+        // Convert to std::string
+        candidates.emplace_back(candidateSpan.data(), candidateSpan.size());
+    }
+
+    WebRTCSessionStruct * existingSession = FindSession(sessionId);
+
+    if (!existingSession)
+    {
+        // If WebRTCSessionID is not null and does not match a value in current sessions.
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Protocols::InteractionModel::Status::ConstraintError);
```

Comment:
should we log on these? It seems useful when I see a constraintError to know what cause was it (e.g. no existing session vs wrong peer/fabric).